### ISSUE TITLE
GradleRunner.withDebug(true) makes test fail on Kotlin build script tests

### ISSUE THREAD

**Issue by jnizet** [gradle/gradle#6862]
Testing gradle build scripts using the Kotlin DSL with the TestKit gradle runner works fine unless `withDebug(true)` is called on the `GradleRunner`. 

### Expected Behavior
The test should pass

### Current Behavior
The tests fail with the following stack trace:

```
org.gradle.testkit.runner.UnexpectedBuildFailure: Unexpected build execution failure in /Users/jb/tmp/gradle-runner-bug/gradle-build with arguments [foo]

Output:

FAILURE: Build failed with an exception.

* What went wrong:
java.lang.ExceptionInInitializerError (no error message)
> Type org.gradle.api.tasks.SourceSet not present

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 1s

	at org.gradle.testkit.runner.internal.DefaultGradleRunner$1.execute(DefaultGradleRunner.java:237)
	at org.gradle.testkit.runner.internal.DefaultGradleRunner$1.execute(DefaultGradleRunner.java:234)
	at org.gradle.testkit.runner.internal.DefaultGradleRunner.run(DefaultGradleRunner.java:298)
	at org.gradle.testkit.runner.internal.DefaultGradleRunner.build(DefaultGradleRunner.java:234)
	at bug.TestKitBugTest.failsButShouldPass(TestKitBugTest.java:20)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.runTestClass(JUnitTestClassExecutor.java:106)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:58)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:38)
	at org.gradle.api.internal.tasks.testing.junit.AbstractJUnitTestClassProcessor.processTestClass(AbstractJUnitTestClassProcessor.java:66)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.processTestClass(SuiteTestClassProcessor.java:51)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:32)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:93)
	at com.sun.proxy.$Proxy2.processTestClass(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.processTestClass(TestWorker.java:117)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:155)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:137)
	at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:404)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:46)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
	at java.lang.Thread.run(Thread.java:748)
```

### Context
I was initially trying to translate the Spring Boot gradle plugins samples to the Kotlin DSL, which are tested using the TestKit.

### Steps to Reproduce (for bugs)
A complete minimal project is available at https://github.com/jnizet/gradle-runner-bug.
Run `./gradlew test`. 

Here's the JUnit test:

```
package bug;

import static org.junit.Assert.assertTrue;

import java.io.File;

import org.gradle.testkit.runner.GradleRunner;
import org.junit.Test;

public class TestKitBugTest {
    @Test
    public void passes() {
        GradleRunner runner = createGradleRunner();
        assertTrue(runner.build().getOutput().contains("hello"));
    }

    @Test
    public void failsButShouldPass() {
        GradleRunner runner = createGradleRunner().withDebug(true);
        assertTrue(runner.build().getOutput().contains("hello"));
    }

    private GradleRunner createGradleRunner() {
        return GradleRunner.create().withProjectDir(new File("gradle-build")).withArguments("foo");
    }
}
```

### Your Environment
```

------------------------------------------------------------
Gradle 4.10.2
------------------------------------------------------------

Build time:   2018-09-19 18:10:15 UTC
Revision:     b4d8d5d170bb4ba516e88d7fe5647e2323d791dd

Kotlin DSL:   1.0-rc-6
Kotlin:       1.2.61
Groovy:       2.4.15
Ant:          Apache Ant(TM) version 1.9.11 compiled on March 23 2018
JVM:          1.8.0_152 (Oracle Corporation 25.152-b16)
OS:           Mac OS X 10.13.6 x86_64
```

**Comment by DPUkyle** [comment]
+1

I also hit this when trying to debug spotbugs/spotbugs-gradle-plugin#61

**Comment by marcoferrer** [comment]
+1 
Same here. Had to disable in the mean time. [ProtobufKotlinDslPluginTest.groovy](https://github.com/google/protobuf-gradle-plugin/blob/548171d87a71c28c7ce3640eb3262295b0c743a3/src/test/groovy/com/google/protobuf/gradle/ProtobufKotlinDslPluginTest.groovy#L27)

**Comment by eskatos** [comment]
Thanks for the report and reproducer @jnizet!

Here are scans with the failure stack trace on Gradle [4.10.2](https://scans.gradle.com/s/gejth2mtj42se/tests/si6uhuz4tdufy-gesgrsmkzeuka) and [5.0-rc-1](https://scans.gradle.com/s/rnq57ho2wwqzc/tests/si6uhuz4tdufy-gesgrsmkzeuka).

```
Caused by: java.lang.TypeNotPresentException: Type org.gradle.api.tasks.SourceSet not present
	at org.gradle.api.reflect.TypeOf.captureTypeArgument(TypeOf.java:280)
	at org.gradle.api.reflect.TypeOf.<init>(TypeOf.java:97)
	at org.gradle.kotlin.dsl.provider.plugins.DefaultProjectSchemaProviderKt$$special$$inlined$typeOf$3.<init>(DefaultProjectSchemaProvider.kt:172)
	at org.gradle.kotlin.dsl.provider.plugins.DefaultProjectSchemaProviderKt.<clinit>(DefaultProjectSchemaProvider.kt:198)
```

It looks like TestKit embedded mode (for debugging) is causing classloading issues.

**Comment by jbartok** [comment]
Problem seems to have been solved in `Gradle 5.3`. Pls. update to that version of higher.
